dev/AppNotifications/AppNotificationUtility.cpp

```diff
@@ -204,31 +210,79 @@ void Microsoft::Windows::AppNotifications::Helpers::RegisterAssets(std::wstring
         &hKey,
         nullptr /* lpdwDisposition */));
 
+    // Retrieve DisplayName and IconUri
+    // 1. We call into Shell APIs, to get both from the app taskbar shortcut.
+    // 2. If 1 doesn't work (i.e. possibly because the developer didn't specify any DisplayName/Icon,
+    //    we take the file name as app name, and let Shell to put a default icon
+    winrt::com_ptr<IPropertyStore> propertyStore;
+    wil::unique_hwnd hWindow{ GetConsoleWindow() };
+    THROW_IF_FAILED(SHGetPropertyStoreForWindow(hWindow.get(), IID_PPV_ARGS(propertyStore.put())));
+
+    // Retrieve the display name
+    std::wstring displayName{};
+
+    wil::unique_prop_variant propVariantDisplayName;
+    THROW_IF_FAILED(propertyStore->GetValue(PKEY_AppUserModel_RelaunchDisplayNameResource, &propVariantDisplayName));
+
+    displayName = propVariantDisplayName.pwszVal != nullptr ? propVariantDisplayName.pwszVal : L"";
+
+    if (displayName.length() == 0)
+    {
+        wil::unique_cotaskmem_string processName;
+        THROW_IF_FAILED(GetCurrentProcessPath(processName));
+
+        displayName = processName.get();
+
+        size_t lastBackslashPosition{ displayName.rfind(L"\\") };
+        THROW_HR_IF(E_UNEXPECTED, lastBackslashPosition == std::wstring::npos);
+        displayName = displayName.substr(lastBackslashPosition + 1); // One after the delimiter
+
+        displayName.erase(displayName.find_first_of(L".")); // Remove file extension
+    }
+
+    // Retrieve the icon
+    std::wstring iconFilePath{};
+
+    wil::unique_prop_variant propVariantIcon;
+    THROW_IF_FAILED(propertyStore->GetValue(PKEY_AppUserModel_RelaunchIconResource, &propVariantIcon));
```

Comment:
THROW_IF_FAILED_MSG(propertyStore->GetValue(PKEY_AppUserModel_RelaunchIconResource, &propVariantIcon), "Couldn't find the Icon Uri Path! Please specifiy it in API blah blah")

Reply:
`GetValue` returns S_OK even if there is no element there. But we can check the propVariant to verify there is no element: https://docs.microsoft.com/en-us/previous-versions/windows/desktop/legacy/bb761473(v=vs.85)#remarks

Reply:
Thats a weird API. But yes, please check propVariant in that case using recommended pattern in msdn.